### DISCUSSION TITLE
How to make change password page accessible only to authenticated users.

### DISCUSSION THREAD

**Discussion by AbdelrahmanAhmed605** [supabase/supabase#20888] [Questions] [answered]
I have a next.js application and I am following the documentation for allowing users to reset passwords.

So far, I built the password reset page and made it publicly accessible with a form asking for the user's email address. Upon submission of the form, I use the supabase.auth.resetPasswordForEmail API to request a password reset link for the user. I also provided the redirectTo parameter pointing to the change password page.

```
await supabase.auth.resetPasswordForEmail(email, {
      redirectTo: `${requestUrl.origin}/change-pass`,
    }
```

Now in the documentation, it says to make the change password page accessible only to authenticated users. In my next.js middleware file I tried:

```
export async function middleware(req: NextRequest) {
  const res = NextResponse.next();

const supabase = createMiddlewareClient({ req, res });

  const {
    data: { user },
  } = await supabase.auth.getUser();

if (!user && req.nextUrl.pathname.startsWith('/update-password')) {
return NextResponse.redirect(new URL('/sign-in', req.url));
}

return res;
```

The idea is that only authenticated users will be allowed to the /update-password page and unauthenticated users will be redirected to the sign-in page. However, right now everytime I try to direct to `req.nextUrl.pathname.startsWith('/update-password')` from the email link sent by supabase, the `supabase.auth.getUser()` and `supabase.auth.getSession()` are both null.

So if both of those are returning null, how else am I supposed to make the change password page accessible only to authenticated users?

**Comment by AbdelrahmanAhmed605** [comment]
I even tried to use the token provided in the link sent from supabase but that did not work. This is what I did:
```
if (req.nextUrl.pathname.startsWith('/update-password')) {
    const requestUrl = new URL(req.url);
    const requestUrl_code = requestUrl.searchParams.get('code');
    const exchange = await supabase.auth.exchangeCodeForSession(
      requestUrl_code
    );
```

but I get the error:
```
⨯ ../node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js (40:10) @ handleError
⨯ invalid request: both auth code and code verifier should be non-empty
```

Is this not the correct approach that the documentation meant when they said make the page accessible only to authenticated users? what approach should I use?

**Comment by charislam** [comment]
Hey, it looks like you're using Next.js with the App Router and Server-Side Auth? (Correct me if I'm wrong.)

I'm curious which instructions you're following (do you have a link?) because they don't seem quite correct. [This page](https://supabase.com/docs/guides/auth/auth-password-reset#server-side-rendering-ssr) explains it quite well, I think.

The reset password is a _two-step_ flow, which your code seems to be conflating into one step.

1. The user is first directed to `/api/auth/callback`, where the code from their email link is exchanged for an actual refresh/access token pair. This is why `getUser` is returning null, because the code is never exchanged, so there's no token. (BTW I see you tried `getSession` -- never use `getSession` on the server because it's [unsafe](https://supabase.com/docs/reference/javascript/auth-getsession).)

2. Then, _after_ the code exchange, you redirect them to `/update-password`, and _that_ is the route you should be making accessible only to authenticated users.

**Reply by AbdelrahmanAhmed605** [reply]
Hey @charislam thanks for the reply! So the link you sent is actually the one I'm following but maybe I am understanding it wrong. Where do you make the call to /api/auth/callback?

What I've done so far is I made a password reset page with a form that takes a users email and once submitted performs the following:
```
const { data: resetData, error: resetError } =
    await supabase.auth.resetPasswordForEmail(email, {
      redirectTo: `${requestUrl.origin}/update-password`,
    });
```

Next the user would go on their email and click the link where they will be re-directed to the /update-password page. On this page is where I have the code:
```
const { data: resetData, error: resetError } =
      await supabase.auth.updateUser({
        password: formData.password,
      });
```

But before the user tries accessing the /update-password page, I tried implementing logic to ensure the user is authenticated, but all my attempts in the nextjs middleware have failed:
```
export async function middleware(req: NextRequest) {
  const res = NextResponse.next();
  const supabase = createMiddlewareClient({ req, res });
if (req.nextUrl.pathname.startsWith('/update-password')) {}
}
```

Now on the link you specified, it says that my /update-password page "Should be accessible only to authenticated users".  I take it that I shouldn't just allow anyone to type "/update-password" and have access to the page so that is why I tried implementing my logic in the nextjs middleware before a user tries accessing that endpoint. Is this not the correct approach?

**Reply by charislam** [reply]
Ahh I see the problem. We seem to have a versioning issue on the page, where the version you're seeing is not the version I'm showing 🤦🏼‍♀️. We'll fix that, but in the meantime, what you want to do is something like this: https://supabase.com/docs/guides/auth/server-side/email-based-auth-with-pkce-flow-for-ssr

Those instructions are for signing up but just replace the `signUp` function with `resetPasswordForEmail`, and redirect with `/update-password`, and it should work

**Reply by AbdelrahmanAhmed605** [reply]
It seems that the site you sent uses "@supabase/ssr", but my project still uses the nextjs auth helpers. Is there no way to use the auth helpers to make the `/update-password` page only accessible to authenticated users?

1. It seems that the next.js middleware cannot catch the auth session from the supabase createMiddlewareClient before the user is re-directed to my `/update-password` page. However, when they click on the link in the email and are re-directed to the `/update-password` page; if the user refreshes the page, they seem to be automatically logged in. So it seems that clicking the email link to reset the password automatically logs the users in.
2. Not only are users automatically logged in, but because I can't catch the auth session from the supabase createMiddlewareClient, I can't re-direct unauthenticated users using my next.js middleware. Right now anyone can type `/update-password` and access the page since it cannot differentiate a random user with a user that clicked the email link (since the middleware cant seem to find the session)

**Reply by charislam** [reply, marked as answer]
You should be able to do the exact same thing with `auth-helpers` (just changes how you create the Supabase client, but the rest of the flow is the same).

Basically you just want the `/auth/callback` route to finish the login, so that the `/update-password` route can catch unauthenticated users (and you can do it on the route itself, not in middleware).

**Reply by AbdelrahmanAhmed605** [reply]
I'll give that a try but I think this is exactly what I am looking for. Thank you so much for all your help @charislam ! I really appreicte it :)
